base/compiler/typeutils.jl

```diff
@@ -124,3 +124,19 @@ function _switchtupleunion(t::Vector{Any}, i::Int, tunion::Vector{Any}, @nospeci
     end
     return tunion
 end
+
+# unioncomplexity estimates the number of calls to `tmerge` to obtain the given type by
+# counting the Union instances, taking also into account those hidden in a Tuple
+unioncomplexity(u::Union) = 1 + unioncomplexity(u.a) + unioncomplexity(u.b)
+function unioncomplexity(t::DataType)
+    t.name === Tuple.name || return 0
+    c = 0
+    for ti in t.parameters
+        ci = unioncomplexity(ti)
+        if ci > c
+            c = ci
+        end
+    end
+    return c
+end
+unioncomplexity(@nospecialize(x)) = 0
```

Comment:
What about `UnionAll`? It looks like we should unwrap them in case there are tuples or unions inside.

Reply:
I actually had that but apparently lost it in one of the updates. I wasn't sure whether it was necessary, but have now managed to come up with a test case. I also think it might be a good idea to also recurse down the upper bound, but couldn't come up with a test case. Ideas how to construct one?